cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/DistributedStreamCoordinator.java

```diff
@@ -15,38 +15,198 @@
+  private static final Logger LOG = LoggerFactory.getLogger(DistributedStreamCoordinator.class);
+
+  private static final String STREAMS_COORDINATOR = "streams.coordinator";
 
   private ZKClient zkClient;
 
+  private final DiscoveryServiceClient discoveryServiceClient;
+  private final StreamMetaStore streamMetaStore;
+  private final ResourceCoordinatorClient resourceCoordinatorClient;
+
+  private LeaderElection leaderElection;
+  private ResourceCoordinator resourceCoordinator;
+  private Discoverable handlerDiscoverable;
+  private Cancellable handlerSubscription;
+
   @Inject
-  protected DistributedStreamCoordinator(StreamAdmin streamAdmin) {
+  public DistributedStreamCoordinator(StreamAdmin streamAdmin, ZKClient zkClient,
+                                      DiscoveryServiceClient discoveryServiceClient,
+                                      StreamMetaStore streamMetaStore) {
     super(streamAdmin);
+    this.zkClient = zkClient;
+    this.discoveryServiceClient = discoveryServiceClient;
+    this.streamMetaStore = streamMetaStore;
+    this.resourceCoordinatorClient = new ResourceCoordinatorClient(zkClient);
+    this.handlerDiscoverable = null;
   }
 
-  @Inject(optional = true)
-  void setZkClient(ZKClient zkClient) {
-    // Use optional injection for zk client to make testing easier in case this class is not used.
-    this.zkClient = zkClient;
+  @Override
+  protected void startUp() throws Exception {
+    Preconditions.checkNotNull(handlerDiscoverable, "Stream Handler discoverable has not been set");
+    resourceCoordinatorClient.startAndWait();
+    handlerSubscription = resourceCoordinatorClient.subscribe(handlerDiscoverable.getName(),
+                                                              new StreamsLeaderHandler());
+
+    // Start the resource coordinator that will map Streams to Stream handlers
+    leaderElection = new LeaderElection(
+      zkClient, "/election/" + STREAMS_COORDINATOR, new ElectionHandler() {
+      @Override
+      public void leader() {
+        LOG.info("Became Stream handler leader. Starting resource coordinator.");
+        resourceCoordinator = new ResourceCoordinator(zkClient, discoveryServiceClient,
+                                                             new BalancedAssignmentStrategy());
+        resourceCoordinator.startAndWait();
+
+        try {
+          // Create one requirement for the resource coordinator for all the streams.
+          // One stream is identified by one partition
+          ResourceRequirement.Builder builder = ResourceRequirement.builder(Constants.Service.STREAMS);
+          for (StreamSpecification spec : streamMetaStore.listStreams()) {
+            LOG.debug("Adding {} stream as a resource to the coordinator to manager streams leaders.", spec.getName());
+            builder.addPartition(new ResourceRequirement.Partition(spec.getName(), 1));
+          }
+          resourceCoordinatorClient.submitRequirement(builder.build()).get();
```

Comment:
I have questions related to the last comment you made about that issue:
When a new leader is elected, do we have to create a new requirement, or is the one that the previous leader created still present? If we can pick up the previously created requirement, then we shouldn't have to go through the loop that creates the requirement in case of a new leader election, right?